kafka_utils/kafka_check/commands/under_replicated.py

```diff
@@ -0,0 +1,129 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from __future__ import absolute_import
+
+import argparse
+import json
+
+import requests
+
+from kafka_utils.kafka_check import status_code
+from kafka_utils.kafka_check.commands.command import get_broker_id
+from kafka_utils.kafka_check.commands.command import KafkaCheckCmd
+
+
+class UnderReplicatedCmd(KafkaCheckCmd):
+
+    def build_subparser(self, subparsers):
+        subparser = subparsers.add_parser(
+            'under_replicated',
+            description='Check under replicated partitions for all brokers in cluster.',
+            help='This command will sum all numbers of under replicated partitions from each broker.'
+                 'It will query jolokia port for receive this data.',
+        )
+        subparser.add_argument(
+            '--first-broker-only',
+            action='store_true',
+            help='If this parameter is specified, it will do nothing and succeed on '
+                 'not first brokers from Kafka cluster. Set --broker-id to -1 to read broker-id '
+                 'from --data-path',
+        )
+        subparser.add_argument(
+            '--jolokia-port',
+            type=convert_to_port,
+            default=8778,
+            help='Port to access to jolokia on each broker.',
+        )
+
+        return subparser
+
+    def run_command(self):
+        """Under_replicated command, checks number of under replicated partitions for
+        all brokers in the Kafka cluster."""
+        broker_list = self.zk.get_brokers()
+
+        if self.args.first_broker_only and not check_run_on_first_broker(broker_list, self.args):
+            return status_code.OK, 'Provided broker is not the first in broker-list'
+
+        count = get_under_replicated(broker_list, self.args.jolokia_port)
+
+        if count == 0:
+            return status_code.OK, 'No under replicated partitions'
+        else:
+            msg = "{under_replicated} under replicated partitions".format(under_replicated=count)
+            return status_code.CRITICAL, msg
+
+
+def convert_to_port(string):
+    """Convert string to valid port."""
+    MIN_PORT_NUMBER = 0
+    MAX_PORT_NUMBER = 65535
+    error_msg = 'Unsigned short integer required, {string} given.'.format(string=string)
+    try:
+        value = int(string)
+    except ValueError:
+        raise argparse.ArgumentTypeError(error_msg)
+    if value < MIN_PORT_NUMBER or value > MAX_PORT_NUMBER:
+        raise argparse.ArgumentTypeError(error_msg)
+    return value
+
+
+def check_run_on_first_broker(broker_list, args):
+    """Returns true if the first broker in broker_list the same as in args."""
+    if args.broker_id is None:
+        raise RuntimeError("Broker id is not specified")
+
+    broker_id = args.broker_id if args.broker_id != -1 else get_broker_id(args.data_path)
+
+    return broker_id == broker_list.iterkeys().next()
```

Comment:
This line relies on the fact that the broker_list returned by the zk module is sorted, but it would probably be better to sort it again, so that this function doesn't have to rely on a particular input format.